### ISSUE TITLE
SSL Handshake error doing mutual authentication

### ISSUE THREAD

**Issue by mschiessl** [greenbone/ospd#240]
Hey OSPD team,

some time ago, i openend a ticket on [ospd-openvas ](https://github.com/greenbone/ospd-openvas/issues/175#issue-539148093) to track a connection error between GVMD and the remote listening (IP) ospd-openvas instance.

I think the root cause is within OSPD.
When connecting from GVMD I am getting the following error:
```
Exception happened during processing of request from ('xxx.xxx.xxx.xx', 55500)
Traceback (most recent call last):
  File "/usr/lib/python3.7/socketserver.py", line 650, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python3.7/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python3.7/socketserver.py", line 720, in __init__
    self.handle()
  File "/opt/openvas/lib/python3.7/site-packages/ospd-2.0.0-py3.7.egg/ospd/server.py", line 127, in handle
    self.server.handle_request(self.request, self.client_address)
  File "/opt/openvas/lib/python3.7/site-packages/ospd-2.0.0-py3.7.egg/ospd/server.py", line 167, in handle_request
    self.server.handle_request(request, client_address)
  File "/opt/openvas/lib/python3.7/site-packages/ospd-2.0.0-py3.7.egg/ospd/server.py", line 297, in handle_request
    req_socket = self.tls_context.wrap_socket(request, server_side=True)
  File "/usr/lib/python3.7/ssl.py", line 423, in wrap_socket
    session=session
  File "/usr/lib/python3.7/ssl.py", line 870, in _create
    self.do_handshake()
  File "/usr/lib/python3.7/ssl.py", line 1139, in do_handshake
    self._sslobj.do_handshake()
OSError: [Errno 0] Error
```

Fore a detailed analysis of the issue, please read the issue within [ospd-openvas](https://github.com/greenbone/ospd-openvas/issues/175#issue-539148093)

All details and versions can be found in the above ticket

**Comment by mverboom** [comment]
I've been trying to get gvmd to ospd-openvas communcation working over TLS for a bit now. Every time it failes when using:

```
gvmd --verify-scanner=<SCANID>
Failed to verify scanner.
```

 All my work has been done on Debian 10. Tried all combinations of certificates etc. Today I tried to setup on Debian 9. Doing exactly the same as I have done on Debian 10 it works without a glitch!

After searching for a bit I found that the TLS connection to ospd is done using TLSv1.2 on Debian 9 (Python 3.5) and TLSv1.3 on Debian 10 (Python 3.7). This was the only difference I could find that could be significant.

As a test on Debian 10 I changed the protocol the ospd server uses in ospd/server.py. I changed
`PROTOCOL_TLS`
to
`PROTOCOL_TLSv1_2`
and restarted ospd-openvas. When running the same `gvmd --verify-scanner` it is now ok!

Is this a known issue and should the configuration when using TLSv1.3 be done differently in order to get communcations to work?

Thanks!

Mark

**Comment by sedrakpc** [comment]
> I've been trying to get gvmd to ospd-openvas communcation working over TLS for a bit now. Every time it failes when using:
> 
> ```
> gvmd --verify-scanner=<SCANID>
> Failed to verify scanner.
> ```
> 
> All my work has been done on Debian 10. Tried all combinations of certificates etc. Today I tried to setup on Debian 9. Doing exactly the same as I have done on Debian 10 it works without a glitch!
> 
> After searching for a bit I found that the TLS connection to ospd is done using TLSv1.2 on Debian 9 (Python 3.5) and TLSv1.3 on Debian 10 (Python 3.7). This was the only difference I could find that could be significant.
> 
> As a test on Debian 10 I changed the protocol the ospd server uses in ospd/server.py. I changed
> `PROTOCOL_TLS`
> to
> `PROTOCOL_TLSv1_2`
> and restarted ospd-openvas. When running the same `gvmd --verify-scanner` it is now ok!
> 
> Is this a known issue and should the configuration when using TLSv1.3 be done differently in order to get communcations to work?
> 
> Thanks!
> 
> Mark

Can you plz show the code snippet where you changed it?

**Comment by mverboom** [comment]
diff for ospd/server.py:

```
274,275c274,275
<         if hasattr(ssl, 'PROTOCOL_TLS'):
<             protocol = ssl.PROTOCOL_TLS
---
>         if hasattr(ssl, 'PROTOCOL_TLSv1_2'):
>             protocol = ssl.PROTOCOL_TLSv1_2
```

That at least works around the problem for me on debian 10.

Regards,

Mark

**Comment by jjnicola** [comment]
Fix with https://github.com/greenbone/ospd/pull/284.

**Comment by ramsaut** [comment]
The changes were merged to ospd-2.0 in #284 . However, the issue remains open in ospd-20.08 and on master.

**Comment by jjnicola** [comment]
> The changes were merged to ospd-2.0 in #284 . However, the issue remains open in ospd-20.08 and on master.

Hi @ramsaut,
The reference system for GVM 20.08 is Debian 10, which comes with python 3.7, which has the "PROTOCOL_TLS" attribute. So this change will not be applied in ospd-20.08/master.
